libs/depgraph/README.md

```diff
@@ -0,0 +1,284 @@
+node representing the item.
+
+Here is a simplified example for Items representing Linux interfaces and routes.
+Only `Name()`, `Type()` and `Dependencies()` methods are shown here. For other
+required methods, please see the `Item` interface definition.
+
+```go
+import "github.com/lf-edge/eve/libs/depgraph"
+
+type LinuxInterface struct {
+    name string
+}
+
+func (intf LinuxInterface) Name() string {
+    return intf.name
+}
+
+func (intf LinuxInterface) Type() string {
+    return "Linux interface"
+}
+
+func (intf LinuxInterface) Dependencies() []depgraph.Dependency {
+    // no dependencies
+    return []depgraph.Dependency{}
+}
+
+// Other Item methods for LinuxInterface are omitted.
+
+type LinuxRoute struct {
+    via    string
+    dst    string
+    metric int
+}
+
+func (route LinuxRoute) Name() string {
+    // Both dst and via should be included to uniquely identify the route
+    // among all the routes.
+    return route.dst + " via " + route.via
+}
+
+func (route LinuxRoute) Type() string {
+    return "Linux route"
+}
+
+func (route LinuxRoute) Dependencies() []depgraph.Dependency {
+    return []depgraph.Dependency{
+        {
+            Item: depgraph.RequiredItem{
+                Type: "Linux interface", // can also use LinuxInterface{}.Type()
+                Name: route.via, // can also use LinuxInterface{name: route.via}.Name()
+            },
+            Description: "Route requires outgoing interface to be configured first",
+        },
+    }
+}
+
+// Other Item methods for LinuxRoute are omitted.
+```
+
+Once all needed `Item` implementations are available, you are ready to build a dependency
+graph to model some kind of system state.
+Graph with some initial content is created using `New()`:
+
+```go
+import "github.com/lf-edge/eve/libs/depgraph"
+
+// using LinuxRoute and LinuxInterface defined above
+
+g := depgraph.New(depgraph.InitArgs{
+    Name:        "MyGraph",
+    Description: "This is my example graph",
+    Items:       []Item{
+        LinuxRoute{via: "eth0", dst: "10.10.0.0/16", metric: 10},
+        LinuxRoute{via: "eth0", dst: "192.168.16.0/24", metric: 10},
+        LinuxInterface{name: "eth0"},
+    },
+})
+```
+
+Graph will automatically create one node for each item and will build edges based on
+dependencies of these items. In the example above, there will be edge from both routes
+pointing to the `eth0` interface.
+`InitArgs` can also contain the initial content of subgraphs.
+
+A single Node is manipulated using `Node()`, `PutNode()` and `DelNode()` methods:
+
+```go
+import (
+    "fmt"
+    "github.com/lf-edge/eve/libs/depgraph"
+)
+
+// Add new Linux route.
+// Note that `PutNode()` takes `Node` pointer so that a reference to the graph can be
+// added into the Node (this is needed for `Node.AsGraph()`).
+n := depgraph.Node{Item: LinuxRoute{via: "eth0", dst: "10.20.0.0/16", metric: 100}}
+g.PutNode(&n)
+
+// Update the node.
+n.metric = 50
+g.PutNode(&n)
```

Comment:
Since you added it by referencer `&n`, shouldn't `n.metric` change the item as stored?

Also, based on your example, the `metric` is a property of the `Item`, not of the `Node`.

Reply:
What does `metric` mean, anyways?

Reply:
Oh, yes, the metric is supposed to be an attribute of the example item (`LinuxRoute`).
But the "Node struct" is indeed confusing. Will get rid of it and respond to your comment about nodes below.